### PR TITLE
Update project configuration and python versions

### LINKED ISSUE #112: Update dependencies to allow for pandas 2.x
Latest version of ``pandas`` is 2.1, but ``qats`` requires version <2. This is probably too strict, unless there are backwards incompatible changes affecting the use of ``pandas`` within ``qats``.

Suggested tasks:

- [ ] Check list of changes in [What's new in 2.1.0](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html), particularly the section [Backwards incompatible API changes](https://pandas.pydata.org/docs/whatsnew/v2.1.0.html#backwards-incompatible-api-changes) to identify possible issues in loosening the pandas requirements.
- [ ] Update ``setup.py`` and ``requirements.txt``.
- [ ] Test installation manually for relevant python versions (HOLD - it may be sufficient for the automatic build scripts to be run on github).

### LINKED ISSUE #113: Add support for python 3.11
Python 3.11 is out since October 2022, see [What's New In Python 3.11](https://docs.python.org/3/whatsnew/3.11.html). 
``qats`` should support the latest Python version unless there are good reasons not too (not likely).

- [ ] Update testing to include python 3.11 (see [.github/worklow/test.yml](https://github.com/dnvgl/qats/blob/master/.github/workflows/test.yml)).
- [ ] Check numpy dependency table and consider updating dependencies and requirements.txt (might not be necessary as long as the numpy/scipy/pandas/etc. requirements are loose enough).

### LINKED ISSUE #114: New project setup, build and release workflow
## Background 

The current release workflow is roughly as follows:

1. Merge relevant PRs into master branch.
2. Update CHANGELOG manually.
3. Assign new version by using ``version_bumper.py``, which sets a new git tag according to specified release type (major, minor or micro).
4. Push new git tag to directly onto master branch. This triggers the publish workflow ([.github/workflows/publish.yml](https://github.com/dnvgl/qats/blob/master/.github/workflows/publish.yml)), which uses Twine build and upload a new release to PyPi.
5. Project setup is handled by [setup.py](https://github.com/dnvgl/qats/blob/master/setup.py) and the version is extracted from the git tag by using [setuptools_scm](https://github.com/pypa/setuptools_scm).

However, the use of ``setup.py`` for project configuration is deprecated and not advised for new projects (see e.g. the note on using setup.py at [the end of this section on setuptools.pypa.io](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html), [this note on setuptools.pypa.io](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setup-py) and [Why you shouldn't invoke setup.py directly](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html) ). 

Following [PEP 518](https://peps.python.org/pep-0518/) and [PEP 621](https://peps.python.org/pep-0621/), the new standard way of doing it is by ``pyproject.toml``. A basic introduction is given on https://setuptools.pypa.io/en/latest/userguide/quickstart.html. Various ways of handling dynamic versioning are then available, although no one single solution seems to dominate as far as I can see. Possible options in the project configuration as well as build and release workflow are described below.

### Option 1 : keep current
**Pros:** 

- It has worked well so far.

**Cons:** 

- Further use of setup.py (invoked directly) is highly discouraged by smarter people.
- Updating the config and workflow should allow for easier implementation of future functionality. 
- A slightly more automated or assisted process (less manual steps) would be welcome. 

### Option 2: use pyproject.toml, keep setuptools_scm

It seems possible to replace ``setup.py`` by ``pyproject.toml`` while keeping ``setuptools_scm`` as the handler for dynamic versioning. See the setuptools_scm README file: https://github.com/pypa/setuptools_scm#setuptools_scm. 

**Pros:** 

- We avoid using the deprecated ``setup.py``. 
- We may still use git tags to assign a new version and initiate a new release.

**Cons:**

- The workflow is still largely the same as that described above, including manual steps.

### Option 3: implement "Release Please"

A third and interesting option is to implement the **Release Please** (https://github.com/googleapis/release-please) package and associated github action. According to its README file, _"Release Please automates CHANGELOG generation, the creation of GitHub releases, and version bumps for your projects."_ (using ``pyproject.toml`` for project configuration).

An example of a project using Release Please is [equinor/tagreader-python](https://github.com/equinor/tagreader-python) (although with Poetry instead of Twine, but that shouldn't make too big a difference).

This seems to be a very interesting option.

**Pros:** 

- Setup is based on ``pyproject.toml``, avoiding ``setup.py``.
- Automates workflow to a larger degree by generation of changelog and handling version bumps.

**Cons:**

- Anticipate more work to get it up and running.

## Summary

- ``pyproject.toml`` is now the recommended way of configuring the python project (replacing the use of ``setup.py``).
- A good quickstart guide for using ``pyproject.toml`` is provided at https://setuptools.pypa.io/en/latest/userguide/quickstart.html.
- The ``setuptools_csm`` GitHub page provide a brief but good description of using ``setuptools_scm`` for dynamic versioning with ``pyproject.toml``. See https://github.com/pypa/setuptools_scm#setuptools_scm.
- Continued use of ``setuptools_scm`` (using git tags to assign package version) gives the least change to the existing build and release workflow.
- A more comprehensive option is to implement the Release Please package (and associated GitHub action), hosted and described at https://github.com/googleapis/release-please. Key points then are:
   * Release Please replaces ``setuptools_scm`` for version handling.
   * Instead of extracting the version from git tags, Release Please assigns/updates the version in ``pyproject.toml`` during the release process and at the same time assigns a git tag with this version (so simply put, reversed order compared to ``setuptools_scm``).
   * In addition, Release Please is designed to also handle updates of CHANGELOG.md.
- An example of a project using Release Please is found at [equinor/tagreader-python](https://github.com/equinor/tagreader-python).

## Conclusion 

**Option 2.**

Release Please (option 3) will be re-considered at a later stage if necessary.

## Tasks

- [x] Any other options not listed that above that can handle project configuration by ``pyproject.toml`` and dynamic versions (possibly but not necessarily by using git tags)?
- [x] Conclude on which option to go for.

Depending on the option selected:

- Update config and release workflow.
- Update README file (relevant sections).
- Test it as far possible, first manually then with the github actions (incl. release).

**Note:** Tasks to be completed are now handled in PR #115.

### PR DESCRIPTION
**Pull request by eneelo**
## Summary

- Move from `setup.py` to `pyproject.toml`
- Move to `poetry`
- Add support for python 3.11 and 3.12, drop support for 3.7
- Move to `qt6` and require `pyside6`
- Looser version requirements for dependencies, to avoid causing trouble for packages depending on qats
- Deprecate changelog

## Tasks
Final tasks:
- [x] remove entry point `qats-app`, update `qats app --link-app` to rely on calling `pythonw.exe` 
- [x] `publish.yml`: include installation of poetry-dynamic-versioning plugin (include in `test.yml` as well?)
- [x] poetry dynamic versioning: update `pyproject.toml` and `README.md`

Previous tasks:
- [x] Move from `setup.py` to `pyproject.toml` and `build`. Fixes #114 (partly).
- [x] Update of release "workflow": remove `version_bumper.py`
- [x] "deprecate" changelog (point to releases).
- [x] Move to `poetry`: update `pyproject.toml`, `test.yml`, `publish.yml` 
- [x] `publish.yml`: update python version, fix warning caused by actions/checkout@v2 and actions/setup-python@v2 (see [here for more info](https://github.blog/changelog/2023-06-13-github-actions-all-actions-will-run-on-node16-instead-of-node12-by-default/)).
- [x] Update python versions: drop support for 3.7, add support for 3.11 and 3.12. Fixes #113.
- [x] Loosen version requirements for dependencies (`requirements.txt`). Fixes #112.
- [x] `pyside6` support, to handle `pyside2` requiring python <3.11
- [x] `qats.app.gui.py`: remove dependency to `pkg_resources.get_distribution` 
- [x] fix: gui shuts down at 'Display' when using `pyside2` (fixed by supporting/requiring `pyside6`)
- [x] test python 3.8 with `pyside6` 
- [x] if `pyside6` works with python 3.8-3.12: require `pyside6` and drop `pyside2`
- [x] fix: gui gives warnings like `QString(View)::contains(): called on an invalid QRegularExpression object (pattern is 'MOORING * [1-2')` when '[' is opened but not yet closed in filter text field
- [x] Update `README.md`
- [x] Update docs/getting_started.rst (python-version, describe pyside6/pyqt6 options, etc.)
- [x] Fix warnings listed by `py.test` (most fixed, but no quick-fix for invalid escape strings within ..math:: in docstrings, e.g., gumbl.py: line 845)

### CODE REVIEW

**Comment by eneelo** [comment]
@tovop PR is not yet ready for review. Added you to keep you informed on what's coming.

**Comment by tovop** [comment]
> @tovop PR is not yet ready for review. Added you to keep you informed on what's coming.

Thanks. I commented on #114

File: .gitignore [hunk 1]
```diff
@@ -106,3 +106,8 @@ ENV/
 # misc
 code-snippets/
 
+# local test folders
+tmp*/*
+
+# Poetry
+poetry.lock
```
Comment (tovop) [inline .gitignore:113]: Should end with blank line.

**Review by tovop — APPROVED** [review]
Did not find any notable discrepancies that needs to be resolved.

**Pull request merged**